### PR TITLE
Add regression test for generated package data

### LINKED ISSUE #329: Add some regression tests on generated package files
### Is Your Feature Request Related to a Problem? Please Describe

We don't currently have any regression tests for the files that are generated by the package template. As an example of where this would have been useful, see https://github.com/UCL-ARC/python-tooling/pull/309. I made a mistake that broke a couple of the `pyproject.toml` fields, that would have been more easily spotted if there was a test that compared `pyproject.toml` against a reference/expected file.

### Describe the Solution You'd Like

Add a test that compares `pyproject.toml` to an expected/reference file, which fails if the generated file doesn't match. I would suggest starting with just `pyprojec.toml` to pin down the testing machinery, and then rolling out to other files in a follow up PR.

### Describe Alternatives You've Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dstansby**
This adds a regression test for generated pacakge data. The test folder now has a data folder inside it which contains a reference copy of the generated package. This is then compared at test time to a newly generated package. Any differences are shown after the failing test with a diff that looks like:

```
E           RuntimeError: Non-zero diff between generated files and expected files.
E           Generated files can be found in /Users/dstansby/software/python-tooling/pytest_output/test_package_generation0/cookiecutter-test.
E
E           --- /Users/dstansby/software/python-tooling/pytest_output/test_package_generation0/cookiecutter-test/README.md
E           +++ /Users/dstansby/software/python-tooling/tests/data/test_package_generation/README.md
E           @@ -119,7 +119,7 @@
E            tox -e docs
E            ```
E            
E           -from the root of the repository. The built docs will be written to
E           +from the root of the repository. The built documentation will be written to
E            `site`.
E            
E            Alternatively to build and preview the documentation locally, in a Python

tests/test_package_generation.py:79: RuntimeError
```

To update the test files with an intentional change, when the test fails the test data is updated with the changes that can then be easily comitted (if desired).

Fixes https://github.com/UCL-ARC/python-tooling/issues/329

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
-addopts = [
-    "--color=yes",
-    "--import-mode=importlib",
-    "--verbose",
-]
-testpaths = [
-    "tests",
+[tool.mypy]
+exclude = [
+    'tests\/data\/test_package_generation\/src\/cookiecutter_test\/__init__\.py',
```
Comment (dstansby) [inline pyproject.toml]: Can someone work out how I can get this to work? mypy doesn't seem to be ignoring the file 😢

Comment (paddyroddy) [reply pyproject.toml]: So I encountered this problem yesterday (along with @alessandrofelder) https://github.com/python/mypy/issues/17977. It's not ideal, but maybe we just exclude from `pre-commit`?

**Comment by paddyroddy** [comment]
Are we sure we want this? Every time there's a new change, this test will have to be updated.

**Review by paddyroddy — CHANGES_REQUESTED** [review]
Are we sure we want this? Every time there's a new change, this test will have to be updated.

File: pyproject.toml [hunk 2]
```diff
@@ -1,13 +1,12 @@
...
 ]
 
+[tool.pytest.ini_options]
+addopts = ["--color=yes", "--import-mode=importlib", "--verbose"]
+testpaths = ["tests"]
```
Comment (paddyroddy) [inline pyproject.toml]: This should be in a separate PR

File: pyproject.toml [hunk 3]
```diff
@@ -27,9 +26,7 @@ lint.per-file-ignores = {"*tests*" = [
     "INP001",
     "T201",
 ]}
-lint.select = [
-    "ALL",
-]
+lint.select = ["ALL"]
```
Comment (paddyroddy) [inline pyproject.toml]: Why are you changing this? It was consistent throughout the file

File: pyproject.toml (hunk 2 above)
Comment (paddyroddy) [reply pyproject.toml]: I see what you've done now. This change shouldn't be done here. It's not in keeping with the rest of the file

**Comment by dstansby** [comment]
> Are we sure we want this? Every time there's a new change, this test will have to be updated.

See https://github.com/UCL-ARC/python-tooling/issues/329 for context and motivation.

Definitely appreciate that this adds a bit more work when changing anything in the template, but I think it's worth it to catch potential errors such as in https://github.com/UCL-ARC/python-tooling/pull/309 which slipped through the PR author and reviewer.

I've tried to make the loop here as simple as possible:

1. Run tests
2. If this test fails, it modifies the test data in place
3. Review if the test data is modified as expected. If it is, commit and push the changes.

File: pyproject.toml (hunk 2 above)
Comment (dstansby) [reply pyproject.toml]: Sorry, I had an overzealous linter. Should be fixed now

File: pyproject.toml (hunk 3 above)
Comment (dstansby) [reply pyproject.toml]: Sorry, I had an overzealous linter. Should be fixed now

**Comment by paddyroddy** [comment]
> See #329 for context and motivation.
> 
> Definitely appreciate that this adds a bit more work when changing anything in the template, but I think it's worth it to catch potential errors such as in #309 which slipped through the PR author and reviewer.

Consider me won over, I'd forgotten about #309

**Comment by dstansby** [comment]
Sigh, putting this back as draft until I can fix the test

**Review by paddyroddy — CHANGES_REQUESTED** [review]
Can you fix (by which I mean exclude) the link checking failures?
image

**Comment by paddyroddy** [comment]
The tests are failing again @dstansby. `main` had updated.

**Review by paddyroddy — APPROVED** [review]
I've fixed this @dstansby. Will approve and merge now so it doesn't need fixing again.
